torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -249,22 +249,12 @@ def _shard_parameters(self) -> None:
                 self.numel_padded_per_param.append(0)
                 continue
 
-            # Save the original storage and free it later on.
-            # Since we're modifying the tensor's storage directly,
-            # make sure the tensor is the sole occupant of the storage.
-            assert (
-                p.storage_offset() == 0
-            ), "The tensor is not the sole occupant of the storage."
-            orig_storage = p.storage()
-
-            # Replace p with the relevant shard.
-            local_shard, num_padded = self._get_shard(p)
-            p.set_(local_shard)  # type: ignore[call-overload]
+            # Replace p.data with the relevant shard.
+            orig_data = p.data
+            p.data, num_padded = self._get_shard(p)
             self.numel_padded_per_param.append(num_padded)
+            _free_storage(orig_data)
```

Comment:
`_free_storage` also does a check that the tensor is the sole occupant of the storage, if there's a possibility this won't always be the case, we can keep the original implementation.